# NCES CCD Non-Fiscal Data Cleaning

## 1. District Directory

This section inspects and cleans the NCES CCD 2024–25 District Directory data.

The Directory file provides the district-level identifiers and geographic information that will serve as the foundation for merging Membership and Staff data later.

In [118]:
import pandas as pd
import zipfile
from pathlib import Path

In [119]:
directory_zip = Path(
    "../data/raw/NCES_CC_Non_Fiscal/Directory/NCES_CCD_2024_25_District_Directory.zip"
)

with zipfile.ZipFile(directory_zip, "r") as z:
    print(z.namelist())

['ccd_lea_029_2425_w_1a_073025.csv', 'ccd_lea_029_2425_w_1a_073025.sas7bdat']


In [120]:
directory_csv = "ccd_lea_029_2425_w_1a_073025.csv"

with zipfile.ZipFile(directory_zip, "r") as z:
    with z.open(directory_csv) as f:
        directory_df = pd.read_csv(f, low_memory=False)

print("Shape:", directory_df.shape)
directory_df.head()

Shape: (19630, 58)


,SCHOOL_YEAR,FIPST,STATENAME,ST,LEA_NAME,STATE_AGENCY_NO,UNION,ST_LEAID,LEAID,MSTREET1,...,G_11_OFFERED,G_12_OFFERED,G_13_OFFERED,G_UG_OFFERED,G_AE_OFFERED,GSLO,GSHI,LEVEL,IGOFFERED,OPERATIONAL_SCHOOLS
0,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,1000 Industrial School Road,...,Yes,Yes,No,No,No,KG,12,Other,As reported,0
1,2024-2025,1,ALABAMA,AL,Albertville City,1,NaN,AL-101,100005,8379 US Highway 431,...,Yes,Yes,No,No,No,PK,12,Other,As reported,7
2,2024-2025,1,ALABAMA,AL,Marshall County,1,NaN,AL-048,100006,12380 US Highway 431 S,...,Yes,Yes,No,No,No,PK,12,Other,As reported,17
3,2024-2025,1,ALABAMA,AL,Hoover City,1,NaN,AL-158,100007,2810 Metropolitan Way,...,Yes,Yes,No,No,No,PK,12,Other,As reported,18
4,2024-2025,1,ALABAMA,AL,Madison City,1,NaN,AL-169,100008,211 Celtic Dr,...,Yes,Yes,No,No,No,PK,12,Other,As reported,13


In [121]:
directory_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 19630 entries, 0 to 19629
Data columns (total 58 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   SCHOOL_YEAR          19630 non-null  str    
 1   FIPST                19630 non-null  int64  
 2   STATENAME            19630 non-null  str    
 3   ST                   19630 non-null  str    
 4   LEA_NAME             19630 non-null  str    
 5   STATE_AGENCY_NO      19630 non-null  int64  
 6   UNION                287 non-null    float64
 7   ST_LEAID             19630 non-null  str    
 8   LEAID                19630 non-null  int64  
 9   MSTREET1             19629 non-null  str    
 10  MSTREET2             868 non-null    str    
 11  MSTREET3             5 non-null      str    
 12  MCITY                19629 non-null  str    
 13  MSTATE               19630 non-null  str    
 14  MZIP                 19630 non-null  int64  
 15  MZIP4                12026 non-null  float64
 1

In [122]:
directory_df.columns.tolist()

['SCHOOL_YEAR',
 'FIPST',
 'STATENAME',
 'ST',
 'LEA_NAME',
 'STATE_AGENCY_NO',
 'UNION',
 'ST_LEAID',
 'LEAID',
 'MSTREET1',
 'MSTREET2',
 'MSTREET3',
 'MCITY',
 'MSTATE',
 'MZIP',
 'MZIP4',
 'LSTREET1',
 'LSTREET2',
 'LSTREET3',
 'LCITY',
 'LSTATE',
 'LZIP',
 'LZIP4',
 'PHONE',
 'WEBSITE',
 'SY_STATUS',
 'SY_STATUS_TEXT',
 'UPDATED_STATUS',
 'UPDATED_STATUS_TEXT',
 'EFFECTIVE_DATE',
 'LEA_TYPE',
 'LEA_TYPE_TEXT',
 'OUT_OF_STATE_FLAG',
 'CHARTER_LEA',
 'CHARTER_LEA_TEXT',
 'NOGRADES',
 'G_PK_OFFERED',
 'G_KG_OFFERED',
 'G_1_OFFERED',
 'G_2_OFFERED',
 'G_3_OFFERED',
 'G_4_OFFERED',
 'G_5_OFFERED',
 'G_6_OFFERED',
 'G_7_OFFERED',
 'G_8_OFFERED',
 'G_9_OFFERED',
 'G_10_OFFERED',
 'G_11_OFFERED',
 'G_12_OFFERED',
 'G_13_OFFERED',
 'G_UG_OFFERED',
 'G_AE_OFFERED',
 'GSLO',
 'GSHI',
 'LEVEL',
 'IGOFFERED',
 'OPERATIONAL_SCHOOLS']

In [123]:
directory_df[
    [
        "LEAID",
        "ST_LEAID",
        "ST",
        "STATENAME",
        "LEA_NAME",
        "LEA_TYPE_TEXT",
        "SY_STATUS_TEXT",
        "UPDATED_STATUS_TEXT",
        "LEVEL",
        "OPERATIONAL_SCHOOLS"
    ]
].head(10)

,LEAID,ST_LEAID,ST,STATENAME,LEA_NAME,LEA_TYPE_TEXT,SY_STATUS_TEXT,UPDATED_STATUS_TEXT,LEVEL,OPERATIONAL_SCHOOLS
0,100002,AL-210,AL,ALABAMA,Alabama Youth Services,Regular public school district that is not a c...,Open,Open,Other,0
1,100005,AL-101,AL,ALABAMA,Albertville City,Regular public school district that is not a c...,Open,Open,Other,7
2,100006,AL-048,AL,ALABAMA,Marshall County,Regular public school district that is not a c...,Open,Open,Other,17
3,100007,AL-158,AL,ALABAMA,Hoover City,Regular public school district that is not a c...,Open,Open,Other,18
4,100008,AL-169,AL,ALABAMA,Madison City,Regular public school district that is not a c...,Open,Open,Other,13
5,100009,AL-600,AL,ALABAMA,Al Inst Deaf And Blind,State operated agency,Open,Open,Other,5
6,100010,AL-610,AL,ALABAMA,Al Sch Of Math And Science,State operated agency,Open,Open,High,0
7,100011,AL-167,AL,ALABAMA,Leeds City,Regular public school district that is not a c...,Open,Open,Other,4
8,100012,AL-115,AL,ALABAMA,Boaz City,Regular public school district that is not a c...,Open,Open,Other,5
9,100013,AL-205,AL,ALABAMA,Trussville City,Regular public school district that is not a c...,Open,Open,Other,5


In [124]:
print("Unique LEAID:", directory_df["LEAID"].nunique())
print("Total rows:", len(directory_df))
print("Duplicate LEAID:", directory_df["LEAID"].duplicated().sum())

Unique LEAID: 19630
Total rows: 19630
Duplicate LEAID: 0


In [125]:
print("LEA types:")
display(directory_df["LEA_TYPE_TEXT"].value_counts(dropna=False))

print("\nSchool-year status:")
display(directory_df["SY_STATUS_TEXT"].value_counts(dropna=False))

print("\nUpdated status:")
display(directory_df["UPDATED_STATUS_TEXT"].value_counts(dropna=False))

LEA types:


LEA_TYPE_TEXT
Regular public school district that is not a component of a supervisory union    13363
Independent charter district                                                      4393
Service agency                                                                     691
Specialized public school district                                                 444
Other local education agency                                                       250
State operated agency                                                              198
Regular public school district that is a component of a supervisory union          173
Supervisory union                                                                  114
Federal operated agency                                                              4
Name: count, dtype: int64


School-year status:


SY_STATUS_TEXT
Open                       19281
Closed                       114
New                          101
Inactive                      59
Future                        57
Changed Boundary/Agency       15
Reopened                       2
Added                          1
Name: count, dtype: int64


Updated status:


UPDATED_STATUS_TEXT
Open                       19281
Closed                       114
New                          102
Inactive                      59
Future                        56
Changed Boundary/Agency       15
Reopened                       2
Added                          1
Name: count, dtype: int64

In [126]:
[col for col in directory_df.columns 
 if any(word in col.upper() for word in ["LAT", "LON", "CITY", "ZIP", "COUNTY", "LOCALE"])]

['MCITY', 'MZIP', 'MZIP4', 'LCITY', 'LZIP', 'LZIP4']

In [127]:
[col for col in directory_df.columns 
 if any(word in col.upper() for word in ["CBSA", "CSA", "NECTA", "URBAN", "METRO"])]

[]

In [128]:
geo_cols = [
    "LEAID",
    "LEA_NAME",
    "ST",
    "STATENAME",
    "FIPST",
    "LCITY",
    "LSTATE",
    "LZIP"
]

directory_df[geo_cols].head(10)

,LEAID,LEA_NAME,ST,STATENAME,FIPST,LCITY,LSTATE,LZIP
0,100002,Alabama Youth Services,AL,ALABAMA,1,Mt Meigs,AL,36057
1,100005,Albertville City,AL,ALABAMA,1,Albertville,AL,35950
2,100006,Marshall County,AL,ALABAMA,1,Guntersville,AL,35976
3,100007,Hoover City,AL,ALABAMA,1,Hoover,AL,35243
4,100008,Madison City,AL,ALABAMA,1,Madison,AL,35758
5,100009,Al Inst Deaf And Blind,AL,ALABAMA,1,Talladega,AL,35160
6,100010,Al Sch Of Math And Science,AL,ALABAMA,1,Mobile,AL,36604
7,100011,Leeds City,AL,ALABAMA,1,Leeds,AL,35094
8,100012,Boaz City,AL,ALABAMA,1,Boaz,AL,35957
9,100013,Trussville City,AL,ALABAMA,1,Trussville,AL,35173


In [129]:
directory_keep = [
    "SCHOOL_YEAR",
    "LEAID",
    "ST_LEAID",
    "LEA_NAME",
    "ST",
    "STATENAME",
    "FIPST",
    "LCITY",
    "LSTATE",
    "LZIP",
    "LEA_TYPE_TEXT",
    "SY_STATUS_TEXT",
    "UPDATED_STATUS_TEXT",
    "LEVEL",
    "OPERATIONAL_SCHOOLS"
]

directory_clean = directory_df[directory_keep].copy()

print("Clean directory shape:", directory_clean.shape)
directory_clean.head()

Clean directory shape: (19630, 15)


,SCHOOL_YEAR,LEAID,ST_LEAID,LEA_NAME,ST,STATENAME,FIPST,LCITY,LSTATE,LZIP,LEA_TYPE_TEXT,SY_STATUS_TEXT,UPDATED_STATUS_TEXT,LEVEL,OPERATIONAL_SCHOOLS
0,2024-2025,100002,AL-210,Alabama Youth Services,AL,ALABAMA,1,Mt Meigs,AL,36057,Regular public school district that is not a c...,Open,Open,Other,0
1,2024-2025,100005,AL-101,Albertville City,AL,ALABAMA,1,Albertville,AL,35950,Regular public school district that is not a c...,Open,Open,Other,7
2,2024-2025,100006,AL-048,Marshall County,AL,ALABAMA,1,Guntersville,AL,35976,Regular public school district that is not a c...,Open,Open,Other,17
3,2024-2025,100007,AL-158,Hoover City,AL,ALABAMA,1,Hoover,AL,35243,Regular public school district that is not a c...,Open,Open,Other,18
4,2024-2025,100008,AL-169,Madison City,AL,ALABAMA,1,Madison,AL,35758,Regular public school district that is not a c...,Open,Open,Other,13


In [130]:
directory_clean.isna().sum()

SCHOOL_YEAR            0
LEAID                  0
ST_LEAID               0
LEA_NAME               0
ST                     0
STATENAME              0
FIPST                  0
LCITY                  1
LSTATE                 0
LZIP                   0
LEA_TYPE_TEXT          0
SY_STATUS_TEXT         0
UPDATED_STATUS_TEXT    0
LEVEL                  0
OPERATIONAL_SCHOOLS    0
dtype: int64

### Directory Cleaning Summary

- Loaded the NCES CCD 2024–25 District Directory file.
- Confirmed 19,630 unique LEAs with no duplicate `LEAID` values.
- Retained district identifiers, state information, location fields, district type, operational status, level, and number of operational schools.
- No latitude/longitude, county, locale, or metropolitan classification fields were present in this file.
- No LEA types were filtered at this stage so that the Directory, Membership, and Staff files can be compared and merged consistently.

In [131]:
membership_zip = Path(
    "../data/raw/NCES_CC_Non_Fiscal/Membership/NCES_CCD_2024_25_District_Membership.zip"
)

with zipfile.ZipFile(membership_zip, "r") as z:
    print(z.namelist())

['ccd_lea_052_2425_l_1a_073025.csv', 'ccd_lea_052_2425_l_1a_073025.sas7bdat']


In [132]:
with zipfile.ZipFile(membership_zip, "r") as z:
    membership_csv = [name for name in z.namelist() if name.endswith(".csv")][0]

    print("Loading:", membership_csv)

    with z.open(membership_csv) as f:
        membership_df = pd.read_csv(f, low_memory=False)

print("Shape:", membership_df.shape)
membership_df.head()

Loading: ccd_lea_052_2425_l_1a_073025.csv
Shape: (3822460, 15)


,SCHOOL_YEAR,FIPST,STATENAME,ST,LEA_NAME,STATE_AGENCY_NO,UNION,ST_LEAID,LEAID,GRADE,RACE_ETHNICITY,SEX,STUDENT_COUNT,TOTAL_INDICATOR,DMS_FLAG
0,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,American Indian or Alaska Native,Female,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
1,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,American Indian or Alaska Native,Male,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
2,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,Asian,Female,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
3,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,Asian,Male,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
4,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,Black or African American,Female,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing


In [133]:
membership_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3822460 entries, 0 to 3822459
Data columns (total 15 columns):
 #   Column           Dtype  
---  ------           -----  
 0   SCHOOL_YEAR      str    
 1   FIPST            int64  
 2   STATENAME        str    
 3   ST               str    
 4   LEA_NAME         str    
 5   STATE_AGENCY_NO  int64  
 6   UNION            float64
 7   ST_LEAID         str    
 8   LEAID            int64  
 9   GRADE            str    
 10  RACE_ETHNICITY   str    
 11  SEX              str    
 12  STUDENT_COUNT    float64
 13  TOTAL_INDICATOR  str    
 14  DMS_FLAG         str    
dtypes: float64(2), int64(3), str(10)
memory usage: 437.4 MB


In [134]:
for i, col in enumerate(membership_df.columns):
    print(i, col)

0 SCHOOL_YEAR
1 FIPST
2 STATENAME
3 ST
4 LEA_NAME
5 STATE_AGENCY_NO
6 UNION
7 ST_LEAID
8 LEAID
9 GRADE
10 RACE_ETHNICITY
11 SEX
12 STUDENT_COUNT
13 TOTAL_INDICATOR
14 DMS_FLAG


In [135]:
directory_output = Path(
    "../data/cleaned/NCES_CCD_2024_25_District_Directory_clean.csv"
)

directory_clean.to_csv(directory_output, index=False)

print("Saved:", directory_output)
print("Shape:", directory_clean.shape)

Saved: ..\data\cleaned\NCES_CCD_2024_25_District_Directory_clean.csv
Shape: (19630, 15)


# Membership Data Inspection

The NCES CCD 2024–25 District Membership file contains district-level student enrollment data in long format.

Each LEA appears across multiple rows representing combinations of:

- Grade
- Race/ethnicity
- Sex

`STUDENT_COUNT` contains the enrollment count for each combination.

Before cleaning or aggregating the Membership data, the category values for `GRADE`, `RACE_ETHNICITY`, `SEX`, `TOTAL_INDICATOR`, and `DMS_FLAG` are inspected to determine how NCES represents subgroup counts and district totals.

The number of unique `LEAID` values is also compared with the cleaned District Directory to confirm how closely the two files align before merging.

In [136]:
print("Grade values:")
print(membership_df["GRADE"].value_counts(dropna=False))

print("\nRace/ethnicity values:")
print(membership_df["RACE_ETHNICITY"].value_counts(dropna=False))

print("\nSex values:")
print(membership_df["SEX"].value_counts(dropna=False))

print("\nTotal indicator values:")
print(membership_df["TOTAL_INDICATOR"].value_counts(dropna=False))

print("\nDMS flag values:")
print(membership_df["DMS_FLAG"].value_counts(dropna=False))

Grade values:
GRADE
No Category Codes    315316
Grade 6              263232
Grade 5              261168
Grade 1              260496
Grade 2              260368
Grade 4              260368
Grade 3              260192
Kindergarten         260048
Grade 8              259248
Grade 7              259104
Grade 9              225808
Grade 10             225600
Grade 11             225328
Grade 12             224928
Pre-Kindergarten     184544
Not Specified         37096
Ungraded              30048
Adult Education        6912
Grade 13               2656
Name: count, dtype: int64

Race/ethnicity values:
RACE_ETHNICITY
American Indian or Alaska Native             470852
Asian                                        470852
Black or African American                    470852
Hispanic/Latino                              470852
Native Hawaiian or Other Pacific Islander    470852
Two or more races                            470852
White                                        470852
Not Specified      

In [137]:
print("Unique LEAIDs:", membership_df["LEAID"].nunique())
print("Unique LEAIDs in Directory:", directory_clean["LEAID"].nunique())

Unique LEAIDs: 18548
Unique LEAIDs in Directory: 19630


In [138]:
print("Race/ethnicity categories:")
for value in membership_df["RACE_ETHNICITY"].value_counts(dropna=False).index:
    print(value)

print("\nSex categories:")
for value in membership_df["SEX"].value_counts(dropna=False).index:
    print(value)

print("\nTotal indicator categories:")
for value in membership_df["TOTAL_INDICATOR"].value_counts(dropna=False).index:
    print(value)

print("\nDMS flag categories:")
for value in membership_df["DMS_FLAG"].value_counts(dropna=False).index:
    print(value)

Race/ethnicity categories:
American Indian or Alaska Native
Asian
Black or African American
Hispanic/Latino
Native Hawaiian or Other Pacific Islander
Two or more races
White
Not Specified
No Category Codes

Sex categories:
Female
Male
Not Specified
No Category Codes

Total indicator categories:
Category Set A - By Race/Ethnicity; Sex; Grade
Derived - Subtotal by Race/Ethnicity and Sex minus Adult Education Count
Subtotal 4 - By Grade
Derived - Education Unit Total minus Adult Education Count
Education Unit Total

DMS flag categories:
Reported
Manual adjustment
Derived
Not reported
Missing
Suppressed


In [139]:
missing_membership_ids = set(directory_clean["LEAID"]) - set(membership_df["LEAID"])

directory_clean[
    directory_clean["LEAID"].isin(missing_membership_ids)
]["LEA_TYPE_TEXT"].value_counts()

LEA_TYPE_TEXT
Service agency                                                                   498
Independent charter district                                                     166
Specialized public school district                                               119
Supervisory union                                                                114
Other local education agency                                                      96
Regular public school district that is not a component of a supervisory union     67
State operated agency                                                             22
Name: count, dtype: int64

In [140]:
membership_totals["SEX"].value_counts(dropna=False)

SEX
No Category Codes    18548
Name: count, dtype: int64

In [141]:
print(membership_totals["DMS_FLAG"].value_counts(dropna=False))

DMS_FLAG
Reported        18499
Not reported       31
Suppressed         12
Missing             6
Name: count, dtype: int64


In [142]:
print("Missing STUDENT_COUNT:", membership_totals["STUDENT_COUNT"].isna().sum())
print("Non-missing STUDENT_COUNT:", membership_totals["STUDENT_COUNT"].notna().sum())

Missing STUDENT_COUNT: 49
Non-missing STUDENT_COUNT: 18499


In [143]:
membership_totals.groupby("DMS_FLAG")["STUDENT_COUNT"].agg(
    ["count", "min", "max", "mean"]
)

,count,min,max,mean
DMS_FLAG,,,,
Missing,0,NaN,NaN,NaN
Not reported,0,NaN,NaN,NaN
Reported,18499,0.0,408026.0,2680.688145
Suppressed,0,NaN,NaN,NaN


# Full Membership Cleaning

The cleaned Membership dataset preserves the full NCES long-format structure so enrollment can be analyzed later by LEA, grade, race/ethnicity, sex, or combinations of these dimensions.

The `TOTAL_INDICATOR` field is retained so detailed records, subtotals, and overall Education Unit Totals remain distinguishable.

The `DMS_FLAG` field is retained so reported, derived, suppressed, missing, and not-reported values remain identifiable.

The cleaned Membership dataset is intended to be analysis-ready without returning to the raw source file.

In [144]:
membership_clean = membership_df.copy()

print("Raw membership shape:", membership_df.shape)
print("Clean membership shape:", membership_clean.shape)

membership_clean.head()

Raw membership shape: (3822460, 15)
Clean membership shape: (3822460, 15)


,SCHOOL_YEAR,FIPST,STATENAME,ST,LEA_NAME,STATE_AGENCY_NO,UNION,ST_LEAID,LEAID,GRADE,RACE_ETHNICITY,SEX,STUDENT_COUNT,TOTAL_INDICATOR,DMS_FLAG
0,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,American Indian or Alaska Native,Female,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
1,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,American Indian or Alaska Native,Male,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
2,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,Asian,Female,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
3,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,Asian,Male,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing
4,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Grade 1,Black or African American,Female,NaN,Category Set A - By Race/Ethnicity; Sex; Grade,Missing


In [145]:
print("Rows preserved:", len(membership_clean) == len(membership_df))
print("Columns preserved:", membership_clean.shape[1] == membership_df.shape[1])

print("\nShape:", membership_clean.shape)
print("Unique LEAIDs:", membership_clean["LEAID"].nunique())

print("\nMissing STUDENT_COUNT:", membership_clean["STUDENT_COUNT"].isna().sum())

print("\nDMS flags:")
print(membership_clean["DMS_FLAG"].value_counts(dropna=False))

Rows preserved: True
Columns preserved: True

Shape: (3822460, 15)
Unique LEAIDs: 18548

Missing STUDENT_COUNT: 37915

DMS flags:
DMS_FLAG
Reported             2489766
Manual adjustment     746124
Derived               548655
Not reported           29960
Missing                 4754
Suppressed              3201
Name: count, dtype: int64


In [146]:
membership_clean["LEAID"] = membership_clean["LEAID"].astype(str)

In [147]:
membership_clean["LEAID"].dtype

<StringDtype(storage='python', na_value=nan)>

## Membership Identifier Standardization

`LEAID` was converted to string format to preserve it as an identifier rather than a numeric measure and to support consistent joins with other NCES district-level datasets.

In [148]:
membership_output = Path(
    "../data/cleaned/NCES_CCD_2024_25_District_Membership_clean.csv"
)

membership_clean.to_csv(membership_output, index=False)

print("Saved:", membership_output)
print("Shape:", membership_clean.shape)

Saved: ..\data\cleaned\NCES_CCD_2024_25_District_Membership_clean.csv
Shape: (3822460, 15)


# 3. District Staff

This section inspects and cleans the NCES CCD 2024–25 District Staff data.

The Staff file contains district-level staffing measures by staff category. The cleaned Staff dataset will preserve all analytically useful staffing detail so it can be filtered and related to the Directory and Membership datasets later using `LEAID`.

In [149]:
staff_zip = Path(
    "../data/raw/NCES_CC_Non_Fiscal/Staff/CES_CCD_2024_25_District_Staff.zip"
)

with zipfile.ZipFile(staff_zip, "r") as z:
    print(z.namelist())

['ccd_lea_059_2425_l_1a_073025.csv', 'ccd_lea_059_2425_l_1a_073025.sas7bdat']


In [150]:
with zipfile.ZipFile(staff_zip, "r") as z:
    staff_csv = [name for name in z.namelist() if name.endswith(".csv")][0]

    print("Loading:", staff_csv)

    with z.open(staff_csv) as f:
        staff_df = pd.read_csv(f, low_memory=False)

print("Shape:", staff_df.shape)
staff_df.head()

Loading: ccd_lea_059_2425_l_1a_073025.csv
Shape: (454247, 13)


,SCHOOL_YEAR,FIPST,STATENAME,ST,LEA_NAME,STATE_AGENCY_NO,UNION,ST_LEAID,LEAID,STAFF,STAFF_COUNT,TOTAL_INDICATOR,DMS_FLAG
0,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,All Other Support Staff,NaN,Category Set A,Missing
1,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Elementary School Counselors,NaN,Category Set A,Missing
2,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Elementary Teachers,NaN,Category Set A,Missing
3,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Instructional Coordinators and Supervisors to ...,NaN,Category Set A,Missing
4,2024-2025,1,ALABAMA,AL,Alabama Youth Services,1,NaN,AL-210,100002,Kindergarten Teachers,NaN,Category Set A,Missing


In [151]:
staff_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 454247 entries, 0 to 454246
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   SCHOOL_YEAR      454247 non-null  str    
 1   FIPST            454247 non-null  int64  
 2   STATENAME        454247 non-null  str    
 3   ST               454247 non-null  str    
 4   LEA_NAME         454247 non-null  str    
 5   STATE_AGENCY_NO  454247 non-null  int64  
 6   UNION            5746 non-null    float64
 7   ST_LEAID         454247 non-null  str    
 8   LEAID            454247 non-null  int64  
 9   STAFF            454247 non-null  str    
 10  STAFF_COUNT      426300 non-null  float64
 11  TOTAL_INDICATOR  454247 non-null  str    
 12  DMS_FLAG         454247 non-null  str    
dtypes: float64(2), int64(3), str(8)
memory usage: 45.1 MB


In [152]:
for i, col in enumerate(staff_df.columns):
    print(i, col)

0 SCHOOL_YEAR
1 FIPST
2 STATENAME
3 ST
4 LEA_NAME
5 STATE_AGENCY_NO
6 UNION
7 ST_LEAID
8 LEAID
9 STAFF
10 STAFF_COUNT
11 TOTAL_INDICATOR
12 DMS_FLAG


### Staff Data Inspection

The NCES CCD 2024–25 District Staff file is stored in long format.

Each LEA appears across multiple rows representing different staff categories. `STAFF_COUNT` contains the staffing measure for each category.

Before cleaning, the available staff categories, total indicators, and data-status flags are inspected so that all analytically useful staffing detail can be preserved in the cleaned dataset.

In [153]:
print("Staff categories:")
for value in staff_df["STAFF"].value_counts(dropna=False).index:
    print(value)

print("\nTotal indicator categories:")
for value in staff_df["TOTAL_INDICATOR"].value_counts(dropna=False).index:
    print(value)

print("\nDMS flag categories:")
for value in staff_df["DMS_FLAG"].value_counts(dropna=False).index:
    print(value)

Staff categories:
All Other Support Staff
LEA Administrative Support Staff
LEA Administrators
LEA Staff
Other Staff
No Category Codes
Guidance Counselors
Teachers
School Staff
School Administrative Support Staff
School administrators
Instructional Coordinators and Supervisors to the Staff
Librarians/media specialists
Library/Media Support Staff
Paraprofessionals/Instructional Aides
School Psychologists
Student Support Services Staff (w/o Psychology)
Elementary Teachers
Elementary School Counselors
Kindergarten Teachers
Secondary Teachers
Secondary School Counselors
Pre-kindergarten Teachers
School Counselors
Ungraded Teachers

Total indicator categories:
Category Set A
Derived - Major Staffing Category
Education Unit Total

DMS flag categories:
Reported
Derived
Not reported
Missing
Suppressed


In [154]:
print("Unique LEAIDs in Staff:", staff_df["LEAID"].nunique())
print("Unique LEAIDs in Directory:", directory_clean["LEAID"].nunique())

print("\nMissing STAFF_COUNT:", staff_df["STAFF_COUNT"].isna().sum())

Unique LEAIDs in Staff: 19406
Unique LEAIDs in Directory: 19630

Missing STAFF_COUNT: 27947


In [155]:
print("Staff categories:")
for value in staff_df["STAFF"].value_counts(dropna=False).index:
    print(value)

print("\nTotal indicator categories:")
for value in staff_df["TOTAL_INDICATOR"].value_counts(dropna=False).index:
    print(value)

print("\nDMS flag categories:")
for value in staff_df["DMS_FLAG"].value_counts(dropna=False).index:
    print(value)

Staff categories:
All Other Support Staff
LEA Administrative Support Staff
LEA Administrators
LEA Staff
Other Staff
No Category Codes
Guidance Counselors
Teachers
School Staff
School Administrative Support Staff
School administrators
Instructional Coordinators and Supervisors to the Staff
Librarians/media specialists
Library/Media Support Staff
Paraprofessionals/Instructional Aides
School Psychologists
Student Support Services Staff (w/o Psychology)
Elementary Teachers
Elementary School Counselors
Kindergarten Teachers
Secondary Teachers
Secondary School Counselors
Pre-kindergarten Teachers
School Counselors
Ungraded Teachers

Total indicator categories:
Category Set A
Derived - Major Staffing Category
Education Unit Total

DMS flag categories:
Reported
Derived
Not reported
Missing
Suppressed


In [156]:
print("Unique LEAIDs in Staff:", staff_df["LEAID"].nunique())
print("Unique LEAIDs in Directory:", directory_clean["LEAID"].nunique())

print("\nMissing STAFF_COUNT:", staff_df["STAFF_COUNT"].isna().sum())

Unique LEAIDs in Staff: 19406
Unique LEAIDs in Directory: 19630

Missing STAFF_COUNT: 27947


In [157]:
missing_staff_ids = set(directory_clean["LEAID"].astype(str)) - set(
    staff_df["LEAID"].astype(str)
)

directory_clean[
    directory_clean["LEAID"].astype(str).isin(missing_staff_ids)
]["LEA_TYPE_TEXT"].value_counts()

LEA_TYPE_TEXT
Independent charter district                                                     165
Regular public school district that is not a component of a supervisory union     35
Other local education agency                                                      19
Service agency                                                                     7
Specialized public school district                                                 3
State operated agency                                                              1
Name: count, dtype: int64

In [158]:
print(staff_df["DMS_FLAG"].value_counts(dropna=False))

DMS_FLAG
Reported        335802
Derived          90498
Not reported     14308
Missing           8287
Suppressed        5352
Name: count, dtype: int64


In [159]:
staff_df.groupby("DMS_FLAG")["STAFF_COUNT"].agg(
    ["count", "min", "max", "mean"]
)

,count,min,max,mean
DMS_FLAG,,,,
Derived,90498,0.0,24389.00,75.740115
Missing,0,NaN,NaN,NaN
Not reported,0,NaN,NaN,NaN
Reported,335802,0.0,55601.49,40.523449
Suppressed,0,NaN,NaN,NaN


In [160]:
staff_clean = staff_df.copy()

staff_clean["LEAID"] = staff_clean["LEAID"].astype(str)

print("Raw staff shape:", staff_df.shape)
print("Clean staff shape:", staff_clean.shape)
print("Rows preserved:", len(staff_clean) == len(staff_df))
print("Columns preserved:", staff_clean.shape[1] == staff_df.shape[1])
print("LEAID dtype:", staff_clean["LEAID"].dtype)

Raw staff shape: (454247, 13)
Clean staff shape: (454247, 13)
Rows preserved: True
Columns preserved: True
LEAID dtype: str


### Full Staff Cleaning

The cleaned Staff dataset preserves the complete NCES long-format staffing structure.

Each LEA may appear across multiple rows representing different staff categories, including teachers, school counselors, school psychologists, student support services staff, administrators, paraprofessionals, and other staffing groups.

All staff categories are retained so staffing measures can be filtered and compared later in Tableau, Power BI, or Python.

`DMS_FLAG` is retained to distinguish reported, derived, suppressed, missing, and not-reported values. Unavailable staffing counts remain missing rather than being converted to zero.

`LEAID` is standardized as a string identifier for consistent relationships with the Directory, Membership, and Fiscal datasets.

In [161]:
staff_output = Path(
    "../data/cleaned/NCES_CCD_2024_25_District_Staff_clean.csv"
)

staff_clean.to_csv(staff_output, index=False)

print("Saved:", staff_output)
print("Shape:", staff_clean.shape)

Saved: ..\data\cleaned\NCES_CCD_2024_25_District_Staff_clean.csv
Shape: (454247, 13)
